### DISCUSSION TITLE
how to use a turbo package as dependency in another package that is dependency of an app?

### DISCUSSION THREAD

**Discussion by rezasohrabi** [vercel/turborepo#8431] [Help] [answered]
### Summary

I hope you're doing well. I'm working on a monorepo setup using Turbo and have encountered an issue with TypeScript .d.ts generation when running my development script. The structure involves using the @repo/icons package inside the @repo/ui package, which is then used as a dependency in apps.

-apps 
    -dashboard
    -landing
-packages
    -ui
    -icons
    -utils
    
When I run the development script using pnpm dev, I receive the following TypeScript error:

    ```javascript
@repo/ui:dev: src/Alert.tsx(4,22): error TS2307: Cannot find module '@repo/icons' or its corresponding type declarations.
@repo/ui:dev: 
@repo/ui:dev: DTS Build failed
@repo/ui:dev: Error: error occured in dts build
@repo/ui:dev:     at Worker.<anonymous> (\node_modules\.pnpm\tsup@8.0.2_postcss@8.4.35_typescript@5.3.3\node_modules\tsup\dist\index.js:2693:26)
@repo/ui:dev:     at Worker.emit (node:events:518:28)
@repo/ui:dev:     at MessagePort.<anonymous> (node:internal/worker:263:53)
@repo/ui:dev:     at [nodejs.internal.kHybridDispatch] (node:internal/event_target:826:20)
@repo/ui:dev:     at exports.emitMessage (node:internal/per_context/messageport:23:28)
@repo/ui:dev: RollupError: Failed to compile. Check the logs above.
@repo/ui:dev:     at Object.error (\node_modules\.pnpm\rollup@4.9.6\node_modules\rollup\dist\shared\parseAst.js:279:30)
@repo/ui:dev:     at Object.error (\node_modules\.pnpm\rollup@4.9.6\node_modules\rollup\dist\shared\rollup.js:820:32)
@repo/ui:dev:     at Object.error (\node_modules\.pnpm\rollup@4.9.6\node_modules\rollup\dist\shared\rollup.js:18688:42)
``` 
    
i use icons in ui pakcage and use ui as dependency in apps, but it gives me typescript dts error. but when i run it because it has generated already error is gone and successfully started,

if I continue to run the project after this error, it starts successfully, implying that the .d.ts files are generated correctly post-error.

@repo/ui Package Configuration
```javascript
{
  "name": "@repo/ui",
  "version": "0.0.0",
  "type": "module",
  "sideEffects": false,
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": "./dist/index.js"
  },
  "scripts": {
    "build": "tsup --config tsup.config.ts",
    "lint": "eslint --fix .",
    "dev": "tsup --config tsup.config.ts --watch",
    "type-check": "tsc --noEmit",
    "clean": "rimraf node_modules && rimraf dist && rimraf .turbo"
  },
  "peerDependencies": {
    "react": "^18.2.0",
    "@dataak/icons": "workspace:*"
  },
  "devDependencies": {
    "@types/node": "^20.12.12",
    "@types/react": "^18.2.61",
    "autoprefixer": "^10.4.18",
    "daisyui": "^4.11.1",
    "eslint-config": "workspace:*",
    "postcss": "^8.4.35",
    "react": "^18.2.0",
    "tailwindcss": "^3.4.1",
    "tsup": "^8.0.2",
    "typescript": "^5.3.3",
    "typescript-config": "workspace:*"
  },
  "dependencies": {
    "clsx": "^2.1.0",
    "react-hook-form": "^7.51.4"
  }
}

``` 

@repo/icons Package Configuration
```javascript
{
  "name": "@repo/icons",
  "version": "0.0.0",
  "type": "module",
  "sideEffects": false,
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": "./dist/index.js"
  },
  "scripts": {
    "build": "tsup --config tsup.config.ts",
    "lint": "eslint --fix .",
    "dev": "tsup --config tsup.config.ts --watch",
    "type-check": "tsc --noEmit",
    "clean": "rimraf node_modules && rimraf dist && rimraf .turbo"
  },
  "peerDependencies": {
    "react": "^18.2.0"
  },
  "devDependencies": {
    "@types/node": "^20.12.12",
    "@types/react": "^18.2.61",
    "esbuild-plugin-svgr": "^2.1.0",
    "eslint-config": "workspace:*",
    "react": "^18.2.0",
    "tsup": "^8.0.2",
    "typescript": "^5.3.3",
    "typescript-config": "workspace:*"
  },
  "dependencies": {
    "clsx": "^2.1.0",
    "react-hook-form": "^7.51.4"
  }
}

``` 

I would really appreciate any guidance or suggestions on how to resolve this issue. It seems the TypeScript declaration files for @repo/icons are not found initially. A potential solution could involve ensuring the @repo/icons package is built and its type declarations are available before running the development server for @repo/ui.

Thank you so much for your help and for maintaining this awesome tool!

**Comment by anthonyshew** [comment, marked as answer]
For context, `tsup` does not generate `.d.ts`files. You'd need `tsc` to do that.

**Reply by rezasohrabi** [reply]
How to configure this scenario using tsup?

I Added the following fields to my ui and icons projects’ tsconfig.json

```
    "outDir": "./dist",
    "noEmit": false,
    "emitDeclarationOnly": true,
    "declaration": true,
    "declarationMap": true
```
    
Modified my build script to use tsup:  `tsup --config tsup.config.ts --onSuccess tsc`


However, it’s not working as expected. Could you provide a real-world example of how to implement this scenario correctly?

**Comment by rezasohrabi** [comment]
For those who have encountered this issue and want to benefit from using tsup, please refer to this template. It resolves the issue by utilizing tsup with TypeScript type declarations: [https://github.com/react18-tools/turborepo-template](https://github.com/react18-tools/turborepo-template).
